huaweicloud/services/eg/data_source_huaweicloud_eg_custom_event_sources.go

```diff
@@ -0,0 +1,173 @@
+							Computed:    true,
+							Description: `The name of the custom event channel to which the custom event source belong.`,
+						},
+						"name": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: `The name of the custom event source.`,
+						},
+						"type": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: `The type of the custom event source.`,
+						},
+						"description": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: `The description of the custom event source.`,
+						},
+						"status": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: `The status of the custom event source.`,
+						},
+						"created_at": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: `The creation time of the custom event source.`,
+						},
+						"updated_at": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: `The update time of the custom event source.`,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func filterEventSources(d *schema.ResourceData, eventSources []custom.Source) ([]interface{}, error) {
+	filter := map[string]interface{}{
+		"ID": d.Get("source_id"),
+	}
+
+	filterResult, err := utils.FilterSliceWithField(eventSources, filter)
+	if err != nil {
+		return nil, fmt.Errorf("error filting list of custom event sources: %s", err)
+	}
+	return filterResult, nil
+}
+
+func flattenCustomEventSources(eventSources []interface{}) []map[string]interface{} {
+	if len(eventSources) < 1 {
+		return nil
+	}
+
+	result := make([]map[string]interface{}, len(eventSources))
+	for i, val := range eventSources {
+		eventSource := val.(custom.Source)
+		result[i] = map[string]interface{}{
+			"id":           eventSource.ID,
+			"channel_id":   eventSource.ChannelId,
+			"channel_name": eventSource.ChannelName,
+			"name":         eventSource.Name,
+			"type":         eventSource.Type,
+			"description":  eventSource.Description,
+			"status":       eventSource.Status,
+			"created_at":   eventSource.CreatedTime,
+			"updated_at":   eventSource.UpdatedTime,
+		}
+	}
+	return result
+}
+
+func dataSourceCustomEventSourcesRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var (
+		cfg    = meta.(*config.Config)
+		region = cfg.GetRegion(d)
+		opts   = custom.ListOpts{
+			ChannelId:    d.Get("channel_id").(string),
+			ProviderType: "CUSTOM",
+			Name:         d.Get("name").(string),
+		}
+	)
+	client, err := cfg.EgV1Client(region)
+	if err != nil {
+		return diag.Errorf("error creating EG v1 client: %s", err)
+	}
+
+	resp, err := custom.List(client, opts)
+	if err != nil {
+		return diag.Errorf("error querying custom event sources: %s", err)
+	}
+	uuid, err := uuid.GenerateUUID()
+	if err != nil {
+		return diag.Errorf("unable to generate ID: %s", err)
+	}
+	d.SetId(uuid)
```

Comment:
it's better to set the data source ID after filtering by ID

Reply:
done.